File: 09 September/09-03/pangram.py
import re

def is_pangram(sentence, letters):
    regex = re.compile(r'[^a-zA-Z]')
    sentence = re.sub(regex, "", sentence).lower()
    
    return is_it_pangram(sentence, letters) and is_it_pangram(letters, sentence)

def is_it_pangram(string1, string2):
    for character in string2:
        if not character in string1:
            return False
    return True

print(is_pangram("hello", "helo"))
print(is_pangram("hello", "hel"))
print(is_pangram("hello", "helow"))
print(is_pangram("hello world", "helowrd"))
print(is_pangram("Hello World!", "helowrd"))
print(is_pangram("Hello World!", "heliowrd"))
print(is_pangram("freeCodeCamp", "frcdmp"))
print(is_pangram("The quick brown fox jumps over the lazy dog.", "abcdefghijklmnopqrstuvwxyz"))

# --- TEST SUITE ---

tests_text = '''
1. is_pangram("hello", "helo") should return True
2. is_pangram("hello", "hel") should return False
3. is_pangram("hello", "helow") should return False
4. is_pangram("hello world", "helowrd") should return True
5. is_pangram("Hello World!", "helowrd") should return True
6. is_pangram("Hello World!", "heliowrd") should return False
7. is_pangram("freeCodeCamp", "frcdmp") should return False
8. is_pangram("The quick brown fox jumps over the lazy dog.", "abcdefghijklmnopqrstuvwxyz") should return True
'''

import re

tests_regex = re.compile(r"(?P<number>\d+)\.\s(?P<function_call>.+) should return (?P<output>.+?)\.?$", re.MULTILINE)
test_data = [item.groupdict() for item in re.finditer(tests_regex, tests_text)]

def run_tests(test_data):
    print("--------------------------")
    print("🧪 Starting Verification...")
    print("--------------------------")

    all_passed = True
    fail_count = 0

    for test in test_data:
        function_call_output = eval(test['function_call'])
        test_output = eval(test['output'])
        
        if function_call_output == test_output:
            print(f"{test['number']}.✅ PASS - Function Call:\n{test['function_call']}")
        else:
            print(f"{test['number']}.❌ FAIL - Function Call:\n{test['function_call']}\nExpected:\n{test_output}\nGot:\n{function_call_output}")
            all_passed = False
            fail_count += 1

    print("----------------------------");

    if (all_passed):
        print("🎉 SUCCESS: All tests PASSED.");
    else:
        print(f"⚠️ WARNING: {fail_count}/{len(test_data)} tests FAILED.");

run_tests(test_data)

File: 09 September/09-03/pangram.js
function isPangram(sentence, letters) {
  sentence = sentence.replace(/[^a-zA-Z]/g, "").toLowerCase();

  return isItPangram(sentence, letters) && isItPangram(letters, sentence);
}

function isItPangram(string1, string2) {
  for (let i = 0; i < string2.length; i ++) {
    if (string1.indexOf(string2[i]) === -1)
      return false;
  }
  return true;
}

console.log(isPangram("hello", "helo"));
console.log(isPangram("hello", "hel"));
console.log(isPangram("hello", "helow"));
console.log(isPangram("hello world", "helowrd"));
console.log(isPangram("Hello World!", "helowrd"));
console.log(isPangram("Hello World!", "heliowrd"));
console.log(isPangram("freeCodeCamp", "frcdmp"));
console.log(isPangram("The quick brown fox jumps over the lazy dog.", "abcdefghijklmnopqrstuvwxyz"));

// --- TEST SUITE ---

const testsText = `
1. isPangram("hello", "helo") should return true
2. isPangram("hello", "hel") should return false
3. isPangram("hello", "helow") should return false
4. isPangram("hello world", "helowrd") should return true
5. isPangram("Hello World!", "helowrd") should return true
6. isPangram("Hello World!", "heliowrd") should return false
7. isPangram("freeCodeCamp", "frcdmp") should return false
8. isPangram("The quick brown fox jumps over the lazy dog.", "abcdefghijklmnopqrstuvwxyz") should return true
`;

const testsRegex = /(?<number>\d+)\.\s(?<functionCall>.+) should return (?<output>.+?)\.?$/gm;
const testData = [...testsText.matchAll(testsRegex).map(match => match.groups)];

function runTests(testData) {
  console.log("--------------------------");
  console.log("🧪Starting Verification...");
  console.log("--------------------------");

  let allPassed = true;
  let failCount = 0;
  
  testData.forEach(test => {
    const functionCallOutput = eval(test.functionCall);
    const testOutput = eval(test.output);
    let comparison;

    if (Array.isArray(testOutput))
      comparison = arraysEqual(functionCallOutput, testOutput);
    else
      comparison = functionCallOutput === testOutput;

    if (comparison) {
      console.log(`${test.number}.✅PASS - Function Call: ${test.functionCall}`);
    }
    else {
      console.log(`${test.number}.❌FAIL - Function Call: ${test.functionCall}\nExpected: ${testOutput}\nGot: ${functionCallOutput}`);
      allPassed = false;
      failCount ++;
    }
  })

  console.log("----------------------------");

  if (allPassed)
    console.log("🎉SUCCESS: All tests PASSED.");
  else
    console.log(`⚠️WARNING: ${failCount}/${testData.length} tests FAILED.`);
}

function arraysEqual(a, b) {
  if (a.length !== b.length)
    return false;

  for (const index in a) {
    if (Array.isArray(a[index])) {
      if (Array.isArray(b[index])) {
        if (!arraysEqual(a[index], b[index]))
          return false;
      }
      else 
        return false;
    }

    else if (a[index] !== b[index])
      return false;
  }

  return true;
}

runTests(testData);